Mission RTL

acfloria opened the pull request:
Implement a RTL which uses the planned mission and has the following characteristics:

1. The mission contains a landing pattern
- If RTL is triggered from the mission mode the mission is continued in the fast forward mode:
    - Jumps are ignored
    - The delay command is ignored
    - Loiter waypoints are converted to simple waypoints
- If RTL is triggered from any other auto mode, the closest waypoint (for FW not a landing WP) is identified and the mission is continued in the fast forward mode from that waypoint.
- If RTL is triggered from a manual mode the vehicle flies directly to the landing waypoint and lands there.
2. The mission does not contain a landing pattern.
- If RTL is triggered the mission is flown in reverse starting from the previous position waypoint to the current mission waypoint.
- If RTL is triggered from a any other auto mode, the closest waypoint is identified and the mission is flown in reverse starting from that waypoint.
- If RTL is triggered from a manual mode the vehicle directly flies to the home position and lands there after the specified time.
- When flying the mission in reverse:
    - Any non-position setpoints are ignored.
    - Any position setpoints are converted to regular waypoints (`NAV_CMD_WAYPOINT`).
    - In case of vtol in multicopter mode a transition to the fw mode is executed before flying the mission in reverse.
3. No valid mission is available
- If no valid mission is available the regular RTL is executed (fly to home and land there)
4. The mission changes while in RTL
- No valid mission was available and a mission with a landing wp is uploaded
    - The closest wp is determined and the mission is continued starting at that wp
- No valid mission was available and a mission without a landing wp is uploaded
    - The closest wp is determined and the mission is flown back starting at that wp
- A mission with a landing wp changes
    - The closest wp is determined and the mission is continued starting at that wp
- A mission without a landing wp changes
    - The closest wp is determined and the mission is flown in reverse starting at that waypoint.
- The mission changes from no landing wp to containing a landing wp
    - The closest non landing wp is determined and the mission is continued from that wp
- The landing wp is removed from the mission
   - The closest mission item is determined and the mission is flown back from that wp.

If the vehicle is in the normal mission mode and a new mission is uploaded then the mission is continued from the closest waypoint to the vehicle at the time the new mission is resumed.

I did test it in SITL for a [multicopter](https://review.px4.io/plot_app?log=000aef49-ca3e-49d7-a67f-7b4b0c394f81), [vtol](https://review.px4.io/plot_app?log=6c8539c9-aafe-4a6f-b442-3e2d4ed50649), and [plane](https://review.px4.io/plot_app?log=49920787-6d82-4d11-a50f-aad112a084f4).
I tried to test and catch all corner cases. However, as it is a safety critical feature it would be good if at least one reviewer would also test it (just set the `RTL_TYPE` parameter to 2).

dagar commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -631,12 +637,46 @@ Navigator::task_main()
-			if (mission_landing_required() && on_mission_landing()) {
-				navigation_mode_new = &_mission;
+			switch (rtl_type()) {
+			case 1:
```
Which should define these somewhere.

Antiheavy replied:
yes.  also "rtl_type" is a better param name than "rtl_land_type" ;-)

dagar commented on src/modules/navigator/mission_reverse.cpp:
```diff
@@ -0,0 +1,399 @@
+}
+
+void
+MissionReverse::set_mission_items()
```
Could we directly call Mission::set_mission_items() instead if the next next position setpoint is first set up?

Antiheavy commented:
Neat extension of https://github.com/PX4/Firmware/pull/8487 - I am interested in testing this on Fixed Wing SITL, mostly to verify that it doesn't break anything with RTL_TYPE=1 since we use it on our vehicles.  It may take me a few days to get to it though.

@acfloria can you give an example use case for this?   Just curious.



> If the mission changes while flying the mission in reverse the closest mission item of the new mission to the current vehicle position is identified. The mission is then flown back starting at that mission item.

Interesting:
This type of functionality should be considered more broadly for uploading a new mission anytime in flight, not just in this case, but that is a separate topic.
Questions:
What if the new mission has a landing pattern?  will the vehicle stop flying the reverse mission and start flying the mission forwards all the way to the landing pattern - or will is continue in reverse?  What would you want it to do?
What if the nearest mission item is a Land waypoint, will it try to land right there?  In the case of Fixed Wing this would be bad as fixed wings require flying a pattern to land safely (This actually touches on a larger issue about Fixed Wing behavior and being able to switch directly to a Land waypoint bypassing the pattern - outside the scope of this PR).



> Only switching from the mission navigation mode is allowed. Switching to RTL from any other mode will result in the execution of the regular RTL.

Not sure this is really what you'd want to happen in all scenarios.  For example, if you are in Hold mode and have a landing pattern in your mission, why not use it for Return similar to RTL_TYPE=1  This would be the preferred situation for many fixed wing applications.

To be honest, Fixed Wing users probably won't want to use this "reverse mission" Return behavior anyway.  They should have a landing pattern and probably want Return mode to come land instead of continuing the mission.

acfloria commented:
> If the mission changes while flying the mission in reverse the closest mission item of the new mission to the current vehicle position is identified. The mission is then flown back starting at that mission item.

I think while in RTL the mission should not be changed, but I just wanted to add some safety feature just in case an inexperienced user changes the mission. Another upload a shorter mission so that the airplane is faster back (for example skip surveys or more direct path to home).

> This type of functionality should be considered more broadly for uploading a new mission anytime in flight, not just in this case, but that is a separate topic.

I agree. I would do it in another PR. As part of the improved switching I basically almost implemented it. I think it just requires a lot of careful testing.

> What if the new mission has a landing pattern? will the vehicle stop flying the reverse mission and start flying the mission forwards all the way to the landing pattern - or will is continue in reverse? What would you want it to do?

In that case it would have switched to regular RTL. I changed it that it in this case would go to the landing pattern. I think it should react to the changed mission as this could be used in a way to still steer the airplane by inserting/removing a landing waypoint.

> What if the nearest mission item is a Land waypoint, will it try to land right there? In the case of Fixed Wing this would be bad as fixed wings require flying a pattern to land safely

Good point, I exclude the landing waypoints from the nearest waypoint search in case of a Fixed Wing.

> Not sure this is really what you'd want to happen in all scenarios. For example, if you are in Hold mode and have a landing pattern in your mission, why not use it for Return similar to RTL_TYPE=1 This would be the preferred situation for many fixed wing applications.

After thinking a bit more about the switching I changed it. Please have a look at the description which I updated.

Antiheavy commented:
> After thinking a bit more about the switching I changed it. Please have a look at the description which I updated.

@acfloria your updated description looks good and seems logical.  I am not qualified to do a code review for you, but I will SITL Fixed Wing test in the coming days and report my findings.

Antiheavy commented:
@acfloria I finally got around to testing my fixed wing scenarios in SITL.  generally it works!  a few notes though:

Tested RTL_TYPE 0, 1, and 2

Tested triggering Return mode while in Mission/Loiter/Hold modes by: QGC button press, Geofence breach, and loss of datalink (simulated)

Everything worked as advertised, but I found a few potential issues:

1) after the vehicle completes flying the mission in reverse, it performs the normal climb to return altitude (RTL_RETURN_ALT), return, and then descend to loiter altitude (RTL_DESCEND_ALT).  I assume this is intended behavior?

2) In all RTL_TYPEs, when climbing to return altitude (RTL_RETURN_ALT) and descending to loiter altitude  (RTL_DESCEND_ALT), the vehicle stops short of reaching the set point altitude by the NAV_FW_ALT_RAD distance (default 10 meters).  In other words, if you are trying to return at 100 meters altitude, it actually returns at 90 meters altitude, and trying to descend to 30 meter loiter will actually loiter at 40 meters.  This seems like a bug in RTL code maybe?

3) In both RTL_TYPE 1 and 2, when in the "mission" phase when Return starts, there are no warning messages or announcements that the vehicle has switched into Return mode and is performing a Return.   For RTL_TYPE 2, the only warning messages that come through are the normal RTL ones that trigger after completing the reverse mission and switching to Home.  For RTL_TYPE 1, there is a good message that prints most of the time "RTL: using mission landing", however, no message is sent/printed/logged if the vehicle happens to already be tracking towards or is already inside the Landing Pattern when Return mode is engaged.  **It is important that those messages are sent down to the ground station as well as logged at the time that Return mode is engaged.** 

playing around with this type of Return mode was kinda fun.  I can see some interesting use cases.  Nice work.

I'm also glad  you implemented a more generic framework for RTL_TYPE as I think there will be more Return flavors coming in the future.

acfloria commented:
@Antiheavy thank you for testing and giving the feedback.

>after the vehicle completes flying the mission in reverse, it performs the normal climb to return altitude (RTL_RETURN_ALT), return, and then descend to loiter altitude (RTL_DESCEND_ALT). I assume this is intended behavior?

Yeah, I choose to stick with this behavior just in case the height of the last waypoint would be set very low. Without a climb the vehicle would fly back and loiter at that altitude which is maybe not safe.

> In all RTL_TYPEs, when climbing to return altitude (RTL_RETURN_ALT) and descending to loiter altitude (RTL_DESCEND_ALT), the vehicle stops short of reaching the set point altitude by the NAV_FW_ALT_RAD distance (default 10 meters). In other words, if you are trying to return at 100 meters altitude, it actually returns at 90 meters altitude, and trying to descend to 30 meter loiter will actually loiter at 40 meters. This seems like a bug in RTL code maybe?

Good catch. The reason was that for the loiter and return waypoints the current altitude of the vehicle instead of the loiter/return altitude is set. I fixed that and now the airplane should track the desired altitude.

> In both RTL_TYPE 1 and 2, when in the "mission" phase when Return starts, there are no warning messages or announcements that the vehicle has switched into Return mode and is performing a Return. For RTL_TYPE 2, the only warning messages that come through are the normal RTL ones that trigger after completing the reverse mission and switching to Home. For RTL_TYPE 1, there is a good message that prints most of the time "RTL: using mission landing", however, no message is sent/printed/logged if the vehicle happens to already be tracking towards or is already inside the Landing Pattern when Return mode is engaged. It is important that those messages are sent down to the ground station as well as logged at the time that Return mode is engaged.

I added info messages when RTL is activated.

acfloria commented:
@dagar can you give it another view?

dagar commented:
This is looking really good. I'd still like to see if dataman access can be solely the responsibility of Mission, and if a few areas can be simplified (navigator_main logic, and activation).

I'll go through it again and make specific suggestions, but overall we can definitely get this in soon.

acfloria commented:
@dagar what is the status of your review?

acfloria commented:
@dagar it passed now the CI. Can you give it a look and merge it if it is good?

dagar commented on src/modules/navigator/navigator.h:
```diff
@@ -298,7 +303,7 @@ class Navigator : public control::SuperBlock
 	vehicle_local_position_s			_local_pos{};		/**< local vehicle position */
 	vehicle_status_s				_vstatus{};		/**< vehicle status */
-
+	vehicle_status_s				_previous_vstatus{}; /**< vehicle status of the previous iteration*/
```
Are you only using the nav state?

dagar replied:
Store only the previous uint8_t nav state.

acfloria replied:
Makes sense. I changed it.

dagar commented:
There are a few more minor things, but I think we can live with this. My main concern is that without comprehensive testing or full documented behaviour it seems like this could be accidentally broken rather easily.

dagar commented on cmake/configs/nuttx_px4fmu-v2_default.cmake:
```diff
@@ -51,14 +56,15 @@ set(config_module_list
 	# distance sensors
 	drivers/distance_sensor/ll40ls
-	drivers/distance_sensor/mb12xx
+	#drivers/distance_sensor/leddar_one
```
Will this fit on current master without the fmu-v2 changes?

dagar commented on src/modules/navigator/mission.cpp:
```diff
@@ -1512,3 +1569,44 @@ Mission::generate_waypoint_from_heading(struct position_setpoint_s *setpoint, fl
+		}
+
+		if (item_contains_position(missionitem)) {
+			// do not consider land waypoints for a fw
```
> do not consider land waypoints for a fw

Why not?

acfloria replied:
The reason is that for a fw we must consider the landing approach. If we would consider the landing waypoint as a closest waypoint we could approach the landing strip with a completely different approach than planned which could result in a crash.

Antiheavy replied:
yep.

dagar commented on src/modules/navigator/mission_reverse.cpp:
```diff
@@ -0,0 +1,284 @@
+}
+
+void
+MissionReverse::command_vtol_transition()
```
This should probably move to MissionBlock as well with a parameter for which transition. It's needed in several places (RTL, Land, Mission, etc).

acfloria replied:
It is differently used in the Mission and Land so I am not sure if it is really helpful to generalize it to MissionBlock.

dagar replied:
I don't see why it shouldn't all be the same. When you use the mission item in a MissionBlock to do the transition we can also wait for acceptance before continuing.

https://github.com/PX4/Firmware/blob/master/src/modules/navigator/mission.cpp#L610-L612
https://github.com/PX4/Firmware/blob/master/src/modules/navigator/mission.cpp#L663-L665

acfloria replied:
Ok, I initially thought of putting the full function to MissionBlock. I now created function which sets the correct values for the mission_item as this is the same for all different use cases.

dagar commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -640,6 +641,11 @@ Navigator::task_main()
 			navigation_mode_new = &_mission;
+
+			if (_navigation_mode == &_mission_reverse) {
+				_mission.switch_from_reverse();
```
This should probably be contained within MisisonReverse inactivation(). It's easier to reason if the regular Mission can be kept in tact at all times except the special case when MissionReverse is actually running.

acfloria replied:
Actually the problem here is the order of execution. As the on_activation() from Mission is executed before on_inactivation of MissionReverse I had to add this flag.

dagar commented on src/modules/navigator/rtl.h:
```diff
@@ -49,6 +49,23 @@ class Navigator;
 class RTL final : public MissionBlock
 {
 public:
+	enum RTLState {
```
Why is RTLState now public?

acfloria replied:
Maybe I used it at some point somewhere else, but changed it again to private now.

dagar commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -652,18 +658,89 @@ Navigator::task_main()
+						navigation_mode_new = &_mission;
+
+					} else {
+						// fly the mission in reverse if switching from a non-manual mode
```
This chunk is hard to follow. Correct me if I'm wrong, but the "switch from nonmission" could be the default MissionReverse on_activation() behaviour. MissionReverse on_inactive() could simply disable if Mission proper is running.

Could you explain the check if already in RTL?

acfloria replied:
- I could make it the default on_activation() behaviour but I prefer not to as the closest mission_item does not necessarily has to be the previous waypoint.

- I also want to reset it in any other mode as you do not know where you are when you changed from RTL to MANUAL and again to RTL

- If you are already in the navigation state AUTO_RTL, you are not in the RTL or manual mode, the vehicle is not landed and the reverse mission is not finished then the mission_reverse mode is activated. In any other case the regular RTL is executed.

acfloria commented:
After some discussion with a co-worker I implement the fast forward mission mode for RTL. In this mode the mission is flown ignoring jumps, loiter waypoints and delay commands to ensure arriving at the landing waypoint faster.

I as well as the co-worker tested all the functionalities.

@dagar any pending changes from your side?

acfloria commented:
I combined the MissionReverse and Mission class based on an input of @dagar .
I redid all the tests for the different switching conditions again to make sure that it still works.

dagar commented on src/modules/navigator/mission.cpp:
```diff
@@ -287,7 +371,9 @@ Mission::find_offboard_land_start()
 		}
 
-		if (missionitem.nav_cmd == NAV_CMD_DO_LAND_START) {
+		if ((missionitem.nav_cmd == NAV_CMD_DO_LAND_START) ||
```
Won't this change the RTL planned mission landing behaviour?

acfloria replied:
I just checked it and using QGC one cannot mess it up because a land start waypoint is required. However, I could also add for the `NAV_CMD_LAND` waypoint the additional requirement that the vehicle needs to be a vtol.

dagar commented on src/modules/navigator/mission.cpp:
```diff
@@ -447,11 +568,13 @@ Mission::set_mission_items()
 
 			if (_navigator->get_land_detected()->landed) {
-				mavlink_log_info(_navigator->get_mavlink_log_pub(), "Mission finished, landed.");
+				mavlink_log_info(_navigator->get_mavlink_log_pub(),
```
Can we log the mission execution direction more explicitly?

acfloria replied:
I added the mission execution type to the mission_result message so it is logged now.

dagar commented on cmake/configs/nuttx_px4fmu-v2_default.cmake:
```diff
@@ -7,7 +7,8 @@ set(config_module_list
 	#
 	#drivers/barometer
-	drivers/differential_pressure
+	#drivers/differential_pressure
```
This can be restored now.

acfloria replied:
Changed

dagar commented:
I still wish we had a better design in place to contain the growing navigator complexity, but I think we can live with this for now.

Let's do some testing to verify everything, especially any potential regressions in and around the navigator <-> rtl interaction. How it changes with respect to landing detected is another concern.

Once we're confident with SITL (both MC + FW) we can ask the test team to fly it.

acfloria commented:
I already did a lot of SITL testing for Fixed-Wing and VTOL. I can do some Multicopter SITL tests tonight.

dagar commented:
@PX4/testflights the first priority to test here is that there aren't any regressions in and around current simple RTL usage. After that we can go over exactly how the other RTL types should function and create appropriate test flight cards.

@hamishwillee would it work well to go over RTL documentation updates now? It might help guide the test team.

dagar commented:
@Antiheavy any additional review and/or testing is always appreciated.

Antiheavy commented:
> @Antiheavy any additional review and/or testing is always appreciated.

I'll take another look tonight.



> @hamishwillee would it work well to go over RTL documentation updates now? It might help guide the test team.

I offered to help with this, but haven't gotten to it.  I think that updating the documentation to reflect the multiple options for RTL_TYPE is now appropriate.

acfloria commented:
I just did SITL tests with the multicopter again and it worked as I expected.

hamishwillee commented on src/modules/navigator/rtl_params.c:
```diff
@@ -105,12 +105,14 @@ PARAM_DEFINE_FLOAT(RTL_LAND_DELAY, -1.0f);
- * @value 1 Planned Landing (Mission)
+ * @value 0 Straight to home position
+ * @value 1 Straight to planned landing (mission)
+ * @value 2 Use mission to fly to a planned landing (if available) or the mission back
```
What is "or the mission back"

hamishwillee commented on src/modules/navigator/rtl_params.c:
```diff
@@ -105,12 +105,14 @@ PARAM_DEFINE_FLOAT(RTL_LAND_DELAY, -1.0f);
 PARAM_DEFINE_FLOAT(RTL_MIN_DIST, 5.0f);
 
 /**
```
1. The docs and QGC call this mode "Return". If possible we should do the same. The point being that not every return is a land - e.g. for FW you return and circle.
2. Value 1 seems odd in the light of value 2 - ie what if value 1 is set but no mission landing point is available? Also what if value 2 is set but person would prefer to land at the starting point. Also, what if the mission has no land command for both these cases - ie is just end of mission or a loiter?

Maybe for this block (assuming value 2 is as I thought. 
```
/**
 * Return type
 *
 * Flight path and destination in the event of Return (RTL) being invoked.
 *
 * @value 0 Return to home position via direct path.
 * @value 1 Return to planned mission landing point.
 * @value 2 Return to planned mission landing point (if available) or reverse mission path and land at take off point.
 * @group Return
 */
```

hamishwillee commented:
>> @hamishwillee would it work well to go over RTL documentation updates now? It might help guide the test team.

> I offered to help with this, but haven't gotten to it. I think that updating the documentation to reflect the multiple options for RTL_TYPE is now appropriate.

As per my comments inline I don't yet understand the RTL_TYPE and how it fits in with "fast forward mode" and exactly what happens in each case. We also need to take care with terminology - RTL is named *Return* in QGC and not every vehicle chooses to "Land" at the end of a return. 

I can see how this might work if triggered in a mission, but it "sounds" in some of the comments like mission RTL might be executed outside of a mission if there is a valid mission on the device. That sounds like a bad idea - I often have old missions on my vehicle that have no relevance to my current manual flying. I presume therefore that this all only applies if you are flying a mission or are on some sort of hold/pause state in a current mission?

@Antiheavy I think that it makes sense to update the initial post in this issue (https://github.com/PX4/Firmware/pull/8749#issue-164587248) so that it is accurate and combines all the other changes. I can then merge this into the appropriate place(s) as discussed in https://github.com/PX4/px4_user_guide/issues/137
Do you think you can create a new post as a response that summarises the current implementation?

SayanR1 commented:
Hello Everyone,

I have tested this (PR#8749) RTL worked without any issues.

1- RTL was tested after testing flight modes: Manual, Position and Stabilized and Mission.

https://logs.px4.io/plot_app?log=1e73f039-fbd4-488c-bd92-c993384d1776 
https://logs.px4.io/plot_app?log=1d67e090-eaaa-4b75-92e8-502d19a5a76f

2- RTL was tested before the mission, take off in position few for a few seconds away from the home point and activated RTL several times without any issues. After that Position, Manual and Stabilized and mission were tested without any issues, RTL was done at the end with no issues.

https://logs.px4.io/plot_app?log=f5d39c8f-0c8b-4926-aec5-a6a52a96e28d

3- RTL was tested at the beginning take off in position and activated RTL without any issues, tested RTL a few more times successfully, tested manual, position and Stabilized did not trigger mission on this flight.

https://logs.px4.io/plot_app?log=11ca9aa9-913c-4b90-8eec-671134ebcff9

Avysuarez commented:
Flights with pixhawk 3 pro (V4PRO) Good flights no issues.
https://review.px4.io/plot_app?log=f420d955-f3ab-41dd-86b8-f9bace610ff2
https://review.px4.io/plot_app?log=751a38df-7c78-4bb9-8b34-74f4b32fac35

dannyfpv commented:
2 flights on the pixhawk mini good flight and rtl on both.
https://review.px4.io/plot_app?log=affce843-ec47-4ce3-ab6c-8f5c62173c8a
https://review.px4.io/plot_app?log=ad4b5cdf-145e-4637-958e-4209bbb6556e

r0gelion commented:
@acfloria Sure, we will be able to test it until next Monday. If you can please provide details of the testing and the expected behavior by each RTL type.

Thank you!

Antiheavy commented:
@acfloria I'm finally getting around to SITL testing this again.  I'll watch for your updates this weekend.  In the meanwhile I'm finding some bad behaviors.  I'll add more if I find other issues.

Flyaway behavior if tracking the Land waypoint in mission mode and Return mode is entered (pressed RTL button, or lost comms, etc)
![image](https://user-images.githubusercontent.com/12117027/38459027-30660eb8-3a6a-11e8-87f2-eee1c53ba45a.png)

Antiheavy commented:
Another issue:

The loiter-to-alt with tangent exit criteria seems messed up for FW.  I don't know if this is caused by this PR or if it is an issue with Master right now.  When going back to the landing loiter in Return mode the tangent exit criteria seems to be ignored and the vehicle tries to land on the path between the center of the loiter and the LandWP, not the pre-planned tangent approach direction.  It seems to happen in Return mode, but not Mission mode.  This is bad for 2 reasons: 1) does not complete the loiter until the heading exit critera, 2) lands on the wrong approach line to landing.
 
![image](https://user-images.githubusercontent.com/12117027/38460299-683fb4f8-3a7c-11e8-87df-eff4cb7ac34b.png)

acfloria commented:
@Antiheavy I also was able to replicate the flyaway behaviour, so I will investigate on how to fix it.

With respect to the other issue: It only happens while in RTL, right? This is expected as Loiter to alt waypoints are ignored. I will check if we allow those waypoints in RTL this issue can be solved

Antiheavy commented:
> With respect to the other issue: It only happens while in RTL, right? This is expected as Loiter to alt waypoints are ignored. I will check if we allow those waypoints in RTL this issue can be solved

Yes, i could only get the bad loiter-to-alt landing pattern behavior to happen while in RTL, both RTL_TYPE 1 and 2

Antiheavy commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -587,18 +589,92 @@ Navigator::run()
+							}
+
+							if (rtl_activated) {
+								mavlink_and_console_log_info(get_mavlink_log_pub(), "RTL Mission activated, fly mission reverse mission");
```
suggest re-wording this to either: "RTL Mission activated, fly reverse mission" or "RTL Mission activated, fly mission in reverse"

acfloria replied:
Changed it.

acfloria commented:
I fixed the fw landing issue. @Antiheavy can you confirm please that it works now?

Antiheavy commented:
> changing the waypoint while flying towards the home position might not have the expected effect.

can this be improved to not cause confusion and/or bad system behavior?  In my experience if there is a way to break the system, users will do so very quickly (...like me!)

Antiheavy commented:
also, I am unable to test this firmware branch with newer versions of QGC (say within the past month or so).  Older QGC versions work fine.  I get this error in newer QGC versions using this firmware:  
![image](https://user-images.githubusercontent.com/12117027/38780948-b2a70d20-40a3-11e8-9dfc-c9c260b8a90e.png)

acfloria commented:
@Antiheavy I managed to fix the issue you described. I redid all the other tests, can you also confirm that it is now working as you expect it?

With respect to your QGC issues I don't know what could be the issue. I tested it with that branch https://github.com/mavlink/qgroundcontrol/pull/6366 and it worked.

Antiheavy commented:
> @Antiheavy I managed to fix the issue you described. I redid all the other tests, can you also confirm that it is now working as you expect it?

@acfloria i just recently got my SITL up and running again.  I will try to test this again this weekend.  In particular, I want to test what happens when flying, then an "infeasible" mission is uploaded, and then Return mode is triggered...

Antiheavy commented:
@acfloria I tested the following version in fixed wing SITL for a couple hours.  The issues above seemed fixed and everything else I tested seemed to behave as I would expect.
![image](https://user-images.githubusercontent.com/12117027/39096013-55d68014-460f-11e8-8bba-2bb5a7720129.png)

It seems like this PR is getting pretty close.  Another possible topic to consider is cleaning up the use of the term "RTL" and move towards using "Return".  There has been discussion on this multiple times in the past.  A large PR like this with changes to Return mode might be an opportunity to make improvements.  thoughts?

acfloria commented:
@PX4/testflights could you test this PR again.

The characteristics of `RTL_TYPE` is the following:

0: Regular RTL, fly directly to the home position

1: If a landing point is available fly directly to the landing waypoint (or the start of the landing sequence). If no landing waypoint is available fly directly to the home position

2: See PR description

r0gelion commented:
@acfloria we use to download the build (px4 file) from jenkins/artifacts, but this last commit (31a533a) couldn't be built by jenkins, could you fix that please?

santiago3dr commented:
few flight on Pixracer (V4)

https://logs.px4.io/plot_app?log=1d825845-190d-4c96-acaf-e06833a53235
RTL_TYPE set to 0 
rtl set yaw, headed to home position; expected behavior

https://logs.px4.io/plot_app?log=bc24916f-5f4c-45c5-bf43-0cb956964f86
RTL_TYPE set to 1; half way through Auto, engaged RTL and copter headed towards pre-established land but very slowly and not in a straight path

https://logs.px4.io/plot_app?log=3fca5a22-ae77-4fc3-97f7-028660d5176f
RTL_TYPE set to 1; never engaged auto; engaged rtl and copter headed towards LAND point very slowly, descending on its way to said point

acfloria commented:
@r0gelion I finally came around fixing the CI Error. Could you now do some test flights?

santiago3dr commented:
couple flights running new commit a70cd4e on Pixracer (V4) with RTL_TYPE set to 2

https://logs.px4.io/plot_app?log=a5362bfd-9859-46c5-838a-e307f298ff38
engaged rtl during the mission, copter did a 180 turn and ran the mission back to its start, successfully landing at home position

https://logs.px4.io/plot_app?log=bb68caef-ab20-43b0-919c-d130dd0ab852
manual take-off; never engaged AUTO; RTL returned to home position as expected

santiago3dr commented:
rtl_type = 1 seems to be working now
https://logs.px4.io/plot_app?log=6537154f-900a-49b8-aa7f-e89d4567e300
https://logs.px4.io/plot_app?log=191414ba-2729-4f95-a0b1-25ac934bd938
https://logs.px4.io/plot_app?log=c97ecfc2-e475-4e27-b9a9-9fcbf1c8a83c

acfloria commented:
@dagar good to go?

dagar commented on msg/mission_result.msg:
```diff
@@ -15,3 +19,5 @@ bool flight_termination		# true if the navigator demands a flight termination fr
 uint16 item_changed_index	# indicate which item has changed
 uint16 item_do_jump_remaining	# set to the number of do jumps remaining for that item
+
+uint8 execution_mode	# indicates the mode in which the mission is executed
```
Is this used?

acfloria replied:
@dagar sorry for the delayed response. This is nowhere directly used. But I added it so that is is logged somewhere and can be used for debugging purpose.

potaito commented:
@acfloria @dagar Question regarding this PR: Is the following behaviour intended?

1. Takeoff
2. Create waypoint mission where some intermediate waypoint is close to the current location
3. Start mission. The vehicle will jump to the closest waypoint and commence the mission from there, skipping all waypoints before that.

Example from SITL test. In this case waypoints 1-4 were skipped:
![qgroundcontrol screenshot](https://user-images.githubusercontent.com/3762382/40966546-152aaa8a-68b1-11e8-9048-7a2719772b83.png)

The changes responsible for this are these two blocks in `mission::on_activation()` and `mission::on_active()`, setting the current waypoint to the closest mission item.

https://github.com/acfloria/Firmware/blob/a70cd4e54567ae774ada1d0197797890c16e717f/src/modules/navigator/mission.cpp#L155-L158

https://github.com/acfloria/Firmware/blob/a70cd4e54567ae774ada1d0197797890c16e717f/src/modules/navigator/mission.cpp#L198-L206

Tested on current master (ca5e22583f9dd81852856a9169523d8135b81db6) with SITL `make posix_sitl_default jmavsim` and default parameters.

acfloria commented:
Yes, it was intended. If a mission is uploaded while flying then the closest mission item is set as the next one. If the mission is uploaded/changed while on ground this check is ignored.

Do you have a strong objection regarding this design?

Antiheavy commented:
Dos this only happen in Return mode or is this the behavior always now, even is mission mode?

MaEtUgR commented:
> If a mission is uploaded while flying then the closest mission item is set as the next one.

We just discussed and this is unintuitive and does not make sense except for some special cases. It should not be default behaviour but rather something the user can enable by checkbox. FYI @julianoes @Stifael

potaito commented:
@Antiheavy It always happens, check the screenshots. It was a common waypoint mission without RTL involved. 

@acfloria 
> Yes, it was intended. If a mission is uploaded while flying then the closest mission item is set as the next one. If the mission is uploaded/changed while on ground this check is ignored.

> Do you have a strong objection regarding this design?

Well, at first I thought it was a bug :p 
When I upload a mission manually, I would expect the drone to fly exactly that mission and not just half of it. I see that it makes great sense when things like RTL are generated as a mission with multiple waypoints. _Then_ there is no need in flying through all waypoints. 

Do you expect people to always start missions from a landed and disarmed state? I  for one never do that and I am probably not the only one.

acfloria commented:
Ok, then I will change it back for the regular mission.

potaito commented:
Splendid, thank you :D

dagar commented:
Please open a new issue so we can fix this for the v1.8.0 ASAP.

Antiheavy commented:
> Ok, then I will change it back for the regular mission.

We should also check the behavior with RTL_LAND=1 after you make your fixes.   Uploading a new mission while in Return mode (RTL_LAND=1) should not change the mode and it should stay flying the landing pattern.

hamishwillee commented:
@Antiheavy Seems to me that we need to get this documentation underway in the coming weeks. Are you still able to do first draft? If not, can you please copy the initial post on behaviour, and create a modified version that describes the final behaviour we ended up with?

potaito commented:
@dagar 
> Please open a new issue so we can change this for the v1.8.0 release ASAP.

Done: #9606

superware commented:
Hello, any news regarding this feature?

potaito commented:
@superware what do you mean with "this feature"?

superware commented:
@potaito "RTL which uses the planned mission" :)

potaito commented:
@superware Mmmmh not sure I can follow. This PR was merged long ago and the feature should be part of PX4 now. But it brought an additional feature with it which we wanted to take out retroactively (issue #9606)

hamishwillee commented:
@potaito In fairness, there are no docs on this. I'm trying to get clarity on exactly how this behaves in https://github.com/PX4/px4_user_guide/issues/565

potaito commented:
Oh yeah absolutely. I'm not pretending to know how this works :sweat_smile: 
But if the feature is not working as it should, @superware should open an issue. That's why I wanted to better understand his question.

hamishwillee commented:
@potaito FYI, see https://github.com/PX4/px4_user_guide/pull/571 for docs :-)
